src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -886,12 +922,54 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
 
                         if (status.ErrorCode == SecurityStatusPalErrorCode.ContextExpired)
                         {
-                            return 0;
+                            receivedEOF = true;
+                            break;
                         }
 
                         throw new IOException(SR.net_io_decrypt, SslStreamPal.GetException(status));
                     }
+
+                    if (_decryptedBytesCount == 0)
+                    {
+                        // This may be service frame to update keys or renegotiation.
+                        // Keep waiting for real data.
+                        if (processedLength > 0)
+                        {
+                            // We have some data. Hand them to caller before going for another iteration.
+                            break;
```

Comment:
Why do we break here? Couldn't there still be more frames to process?

Reply:
added check for presence of more data. Alternatively, I can wrap the decrypt data operation with 
```c#
if (_decryptedBytesCount > 0) 
{
.....
}
```
and have the normal closing conditions to kick in instead of another exist here.
```